include/seqan3/search/dream_index/interleaved_bloom_filter.hpp

```diff
@@ -293,6 +293,59 @@ class interleaved_bloom_filter
         };
     }
 
+    /*!\brief Clears a specific bin.
+     * \param[in] bin The bin index to clear.
+     *
+     * \attention This function is only available for **uncompressed** Interleaved Bloom Filters.
+     *
+     * \details
+     *
+     * ### Example
+     *
+     * \include test/snippet/search/dream_index/interleaved_bloom_filter_clear.cpp
+     */
+    void clear(bin_index const bin) noexcept
+    //!\cond
+        requires (data_layout_mode == data_layout::uncompressed)
+    //!\endcond
+    {
+        assert(bin.get() < bins);
+        for (size_t idx = bin.get(), i = 0; i < bin_size_; idx += technical_bins, ++i)
+            data[idx] = 0;
+    }
+
+    /*!\brief Clears a range of bins.
+     * \tparam rng_t The type of the range. Must model std::ranges::forward_range and the reference type must be
+     *               seqan3::bin_index.
+     * \param[in] bin_range The range of bins to clear.
+     *
+     * \attention This function is only available for **uncompressed** Interleaved Bloom Filters.
+     *
+     * \details
+     *
+     * ### Example
+     *
+     * \include test/snippet/search/dream_index/interleaved_bloom_filter_clear.cpp
+     */
+    template <typename rng_t>
+    //!\cond
+        requires (data_layout_mode == data_layout::uncompressed)
+    //!\endcond
+    void clear(rng_t && bin_range) noexcept
```

Comment:
I guess that this function only exists in the case that bins are close to each other and clearing them would be more cache efficient?

Reply:
But that would assume sorting on the bins. I guess you collect all bins to be cleared in a vector. Do you collect them in-order?

Reply:
> I guess that this function only exists in the case that bins are close to each other and clearing them in bulk would be more cache efficient?

Yes, has a similar effect as having a range overload for `emplace`.
I did some benchmarks, and you have some cache effects, the more bins you have and the bigger the IBF, the better the speedup (inversely, for small IBFs (a few MiB) and low bin count (<256), the non-range overload is faster). I did not look into how many bins you need to clear at once.
I saw x10 for 4 GiB and 1024 bins for clearing all bins.

> But that would assume sorting on the bins. I guess you collect all bins to be cleared in a vector. Do you collect them in-order?

I kinda assume sorting - I don't know how the bins are collected in an application.
I could:
  * Sort the range (I would then require `std::ranges::random_access_range`)
    * If I do this, I might as well make the range unique?
  * Document that it would be good if the range was sorted (and unique).

Reply:
> Yes, has a similar effect as having a range overload for `emplace`.

We don't have one right?

> I kinda assume sorting - I don't know how the bins are collected in an application. I could:

I'm torn because I know that this can have a performance impact. The main problem is that the data layout is in strives and I would assume that most bins (~1024) fit in one cache line and even "few" bins should see some benefit from that even if not sorted (since they fall in the same cache line).

Form an interface perspective a range "for a" clear is somehow strange.

What I'm currently thinking whether we can use the "binning_bitvector" to clear the bins?

Reply:
> We don't have one right?

Not yet, but we should add it soon.

> Form an interface perspective a range "for a" clear is somehow strange.

`erase` would probably be a better name; however `erase` usually also does not have a range interface

> What I'm currently thinking is whether we can use the "binning_bitvector" to clear the bins?

How do you mean exactly?
But now that you mention it, it would probably be better to maybe manipulate words instead of bits.
I.e. erasing bins [1,2,3,63] would be one "word access", but 4 "bit accesses".

Reply:
> How do you mean exactly?

I was asking myself: "When do you clear?" Probably because of a "contain" query. So can't we use the same data type?

Reply:
You want to erase a bin as part of an update procedure.
If you build an IBF over a database and, say, `genome_1` is in `bin_1`, but then in an update `genome_1` gets deleted, you will want to erase `bin_1`.

So basically, you need `erase` to do remove and edit of bins affected by some database update.